torch/sparse/semi_structured.py

```diff
@@ -232,6 +232,33 @@ def __repr__(self) -> str:  # type: ignore[override]
 
     __torch_function__ = torch._C._disabled_torch_function_impl
 
+    def _pad_tensor_for_matmul(self, original_tensor : torch.Tensor) -> torch.Tensor:
+        """
+        Calculates padding for dense tensor and pads tensor if necessary.
+        If padding is not required, this function returns the original tensor.
+        """
+        # only 2d matmul
+        assert original_tensor.dim() == 2
+
+        # check shape
+        m, n = original_tensor.shape
+        min_rows = _DTYPE_TO_SEMI_STRUCTURED_SPARSE_CONFIG[original_tensor.dtype].min_rows
+        min_cols = _DTYPE_TO_SEMI_STRUCTURED_SPARSE_CONFIG[original_tensor.dtype].min_cols
+        to_pad_m = -m % min_rows if m < min_rows or m % min_rows else 0
+        to_pad_n = -n % min_cols if n < min_cols or n % min_rows else 0
+        if to_pad_m or to_pad_n:
+            warnings.warn(
+                (
+                    "Attempting to do matmul with a dense tensor that does not meet shape requirements."
+                    "Padding dense input tensor of shape ({m}, {n}) to ({m+to_pad_m}, {n+to_pad_n}."
+                ),
+                UserWarning,
+            )
+            return torch.nn.functional.pad(original_tensor, (0, to_pad_n, 0, to_pad_m))
```

Comment:
I think (hope) F.pad is smart enough to do nothing when `to_pad` is 0. 

Also, can you try this code with torch.compile to see how the compiler handles the padding?

Reply:
Yes, I will run our benchmarks again too to make sure there is no perf regression. 

For torch.compile we don't have brians subclassing support yet. 
Do you want me to just verify that we can fuse padding with a test? 

I can also just open a new branch with the same padding changes in saf, if you want to try this e2e

Reply:
@jcaip I think Brian's work actually just recently landed (at least I thought it did)

Reply:
Oh perfect, I'll try that then